src/base/bittorrent/session.cpp

```diff
@@ -1891,8 +1893,46 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
 
         p = magnetUri.addTorrentParams();
     }
-    else if (torrentInfo.isValid()) {
-        if (!params.restored) {
+    else {
+        if (!torrentInfo.isValid()) {
+            // We can have an invalid torrentInfo when there isn't a matching
+            // .torrent file to the .fastresume we loaded. Possibly from a
+            // failed upgrade.
+            return false;
+        }
+
+        hash = torrentInfo.hash();
+    }
+
+    // We should not add the torrent if it is already
+    // processed or is pending to add to session
+    if (m_addingTorrents.contains(hash) || m_loadedMetadata.contains(hash))
+        return false;
+
+    TorrentHandle *const torrent = m_torrents.value(hash);
+    if (torrent) {  // a duplicate torrent is added
+        if (torrent->isPrivate() || (!fromMagnetUri && torrentInfo.isPrivate()))
+            return false;
+
+        // merge trackers and web seeds
+        torrent->addTrackers(fromMagnetUri ? magnetUri.trackers() : torrentInfo.trackers());
+        torrent->addUrlSeeds(fromMagnetUri ? magnetUri.urlSeeds() : torrentInfo.urlSeeds());
+        return true;
+    }
+
+    if (!fromMagnetUri) {
+        if (params.restored) {  // load from existing fastresume
+#if (LIBTORRENT_VERSION_NUM < 10200)
+            p.resume_data = std::vector<char> {fastresumeData.constData()
+                , (fastresumeData.constData() + fastresumeData.size())};
+            p.flags |= lt::add_torrent_params::flag_use_resume_save_path;
+#else
+            lt::error_code ec;
+            p = lt::read_resume_data(fastresumeData, ec);
+            Q_ASSERT(!ec);
```

Comment:
Bad idea.
You should handle the case when fastresume data is corrupted (e.g. due to incorrect app termination).

Reply:
>You should handle the case when fastresume data is corrupted (e.g. due to incorrect app termination).

Corrupted or not, `read_resume_data` will always return a valid `add_torrent_params`, so I'm not really sure what kind of error handing is needed here.

Reply:
Besides, even if the fastresume is corrupted we already have all the information so it is still possible to add the torrent to session. This is better than `return false;` here, isn't?

Reply:
So why Q_ASSERT is needed? It's bad idea to assert on something from outer world (I mean fastresume data).

Reply:
>So why Q_ASSERT is needed? 

I assumed it might be useful when in debugging mode. I don't mind removing it if you think it is useless.

Reply:
Failed assertions cause the application crash for me. If you want help to debug it's better to print debug message (qDebug).

Reply:
I've removed it instead.